config/opal_setup_cc.m4

```diff
@@ -47,29 +45,42 @@ AC_DEFUN([OPAL_PROG_CC_C11_HELPER],[
-                   [Whether C compiler support atomic convenience variables in stdatomic.h])
+                   [[#include <stdatomic.h>]], [[static atomic_long foo = 1;++foo;]])
 
     OPAL_CC_HELPER([if $CC $1 supports C11 _Atomic keyword], [opal_prog_cc_c11_helper__Atomic_available],
-                   [[#include <stdatomic.h>]],[[static _Atomic long foo = 1;++foo;]], [OPAL_C_HAVE__ATOMIC],
-                   [Whether C compiler supports __Atomic keyword])
+                   [[#include <stdatomic.h>]],[[static _Atomic long foo = 1;++foo;]])
 
     OPAL_CC_HELPER([if $CC $1 supports C11 _Generic keyword], [opal_prog_cc_c11_helper__Generic_available],
-                   [[#define FOO(x) (_Generic (x, int: 1))]], [[static int x, y; y = FOO(x);]], [OPAL_C_HAVE__GENERIC],
-                   [Whether C compiler supports __Generic keyword])
+                   [[#define FOO(x) (_Generic (x, int: 1))]], [[static int x, y; y = FOO(x);]])
 
     OPAL_CC_HELPER([if $CC $1 supports C11 _Static_assert], [opal_prog_cc_c11_helper__static_assert_available],
-                   [[#include <stdint.h>]],[[_Static_assert(sizeof(int64_t) == 8, "WTH");]], [OPAL_C_HAVE__STATIC_ASSERT],
-                   [Whether C compiler support _Static_assert keyword])
+                   [[#include <stdint.h>]],[[_Static_assert(sizeof(int64_t) == 8, "WTH");]])
+
+    AS_IF([test $opal_prog_cc_c11_helper__Thread_local_available -eq 1 && test $opal_prog_cc_c11_helper_atomic_var_available -eq 1],
+          [$2],
+          [$3])
 
     dnl At this time Open MPI only needs thread local and the atomic convenience types for C11 support. These
     dnl will likely be required in the future.
-    AS_IF([test "x$opal_prog_cc_c11_helper__Thread_local_available" = "xyes" && test "x$opal_prog_cc_c11_helper_atomic_var_available" = "xyes"],
-          [$2], [$3])
+    AS_IF([test "$4" = "yes" ||
+          (test $opal_prog_cc_c11_helper__Thread_local_available -eq 1 && test $opal_prog_cc_c11_helper_atomic_var_available -eq 1)],
+          [AC_DEFINE_UNQUOTED([OPAL_C_HAVE__THREAD_LOCAL], [$opal_prog_cc_c11_helper__Thread_local_available],
+                              [Whether C compiler supports __Thread_local])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE_ATOMIC_CONV_VAR], [$opal_prog_cc_c11_helper_atomic_var_available],
+                              [Whether C compiler support atomic convenience variables in stdatomic.h])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__ATOMIC], [$opal_prog_cc_c11_helper__Atomic_available],
+                              [Whether C compiler supports __Atomic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__GENERIC], [$opal_prog_cc_c11_helper__Generic_available],
+                              [Whether C compiler supports __Generic keyword])
+
+           AC_DEFINE_UNQUOTED([OPAL_C_HAVE__STATIC_ASSERT], [$opal_prog_cc_c11_helper__static_assert_available],
+                              [Whether C compiler support _Static_assert keyword])])
```

Comment:
Why not move these AC_DEFINEs down into OPAL_SETUP_CC?  That way, you don't need a 4th argument and a bogus extra call into OPAL_PROG_CC_C11_HELPER.